File: crates/mithril-oxide-sys/cpp/Dialect/EmitC/IR/EmitC.cpp
#include "EmitC.hpp"

#include <mlir/Dialect/EmitC/IR/EmitC.h>
#include <mlir/IR/MLIRContext.h>
#include <mlir/IR/Types.h>
#include <mlir/Support/LLVM.h>
#include <rust/cxx.h>


namespace mithril_oxide_sys::emitc {

const void *OpaqueType_get(MLIRContext *ctx, rust::Str name)
{
    return OpaqueType::get(ctx, mlir::StringRef(name.data(), name.length())).getAsOpaquePointer();
}

const void *PointerType_get(const void *type)
{
    return PointerType::get(mlir::Type::getFromOpaquePointer(type)).getAsOpaquePointer();
}

} // mithril_oxide_sys::emitc

File: crates/mithril-oxide-sys/src/Dialect/EmitC/IR/EmitC.rs
pub use self::ffi::{OpaqueType_get, PointerType_get};

#[cxx::bridge]
mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/EmitC/IR/EmitC.hpp");

        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        type c_void = crate::c_void;
    }

    #[namespace = "mithril_oxide_sys::emitc"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/EmitC/IR/EmitC.hpp");

        pub unsafe fn OpaqueType_get(context: *mut MLIRContext, name: &str) -> *const c_void;
        #[must_use]
        pub unsafe fn PointerType_get(type_: *const c_void) -> *const c_void;
    }
}
